File: web/projects/portal/src/app/vsobjects/util/property-dialog.service.ts
/*
 * inetsoft-web - StyleBI is a business intelligence web application.
 * Copyright © 2024 InetSoft Technology (info@inetsoft.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
import { Injectable } from "@angular/core";
import { Tool } from "../../../../../shared/util/tool";
import { NgbModal } from "@ng-bootstrap/ng-bootstrap";
import { ModelService } from "../../widget/services/model.service";
import { ComponentTool } from "../../common/util/component-tool";

@Injectable({
    providedIn: "root"
})
export class PropertyDialogService {
   constructor(private modelService: ModelService,
               private modalService: NgbModal)
   {
   }

   public checkScript(id: string, scripts: string[], ok: Function, cancel: Function): void {
      const uri = "../api/composer/vs/check-script/" + Tool.byteEncode(id);
      this.modelService.sendModel<string>(uri, scripts)
         .subscribe(res => {
            const msg = res.body;

            if(msg) {
               this.showError(msg).then(yes => yes ? ok() : cancel());
            }
            else {
               ok();
            }
         });
   }

   private showError(error: string): Promise<boolean> {
      const msg = Tool.formatCatalogString("_#(js:viewer.viewsheet.scriptFailed)", [error]);
      const buttons = {"yes": "_#(js:Yes)", "no": "_#(js:No)"};
      return ComponentTool.showConfirmDialog(this.modalService, "_#(js:Error)", msg, buttons)
         .then(button => Promise.resolve(button == "yes"));
   }
}

File: core/src/main/java/inetsoft/web/composer/vs/dialog/PropertyDialogService.java
/*
 * This file is part of StyleBI.
 * Copyright (C) 2025  InetSoft Technology
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package inetsoft.web.composer.vs.dialog;

import inetsoft.analytic.composition.ViewsheetService;
import inetsoft.cluster.*;
import inetsoft.report.composition.RuntimeViewsheet;
import inetsoft.report.composition.WorksheetEngine;
import inetsoft.web.composer.vs.objects.controller.VSObjectPropertyService;
import inetsoft.web.composer.vs.objects.controller.VSTrapService;
import inetsoft.web.viewsheet.service.VSDialogService;
import org.springframework.stereotype.Service;

import java.security.Principal;

@Service
@ClusterProxy
public class PropertyDialogService {

   public PropertyDialogService(VSObjectPropertyService vsObjectPropertyService,
                                ViewsheetService engine,
                                VSDialogService dialogService,
                                VSTrapService trapService)
   {
      this.vsObjectPropertyService = vsObjectPropertyService;
      this.engine = engine;
      this.dialogService = dialogService;
      this.trapService = trapService;
   }

   @ClusterProxyMethod(WorksheetEngine.CACHE_NAME)
   public String checkScript(@ClusterProxyKey String runtimeId, String[] scripts, Principal principal) throws Exception
   {
      RuntimeViewsheet rvs = engine.getViewsheet(runtimeId, principal);

      for(String script: scripts) {
         String error = vsObjectPropertyService.checkScript(rvs, script);

         if(error != null) {
            return error;
         }
      }

      return null;
   }

   private final ViewsheetService engine;
   private final VSObjectPropertyService vsObjectPropertyService;
   private final VSDialogService dialogService;
   private final VSTrapService trapService;
}
